shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs

```diff
@@ -0,0 +1,110 @@
+import Cardano.Ledger.Era hiding (Crypto)
+import Cardano.Ledger.Shelley (ShelleyEra)
+import Data.Bifunctor (bimap)
+import Data.Functor.Identity (Identity)
+import qualified Data.Map.Strict as Map
+import Shelley.Spec.Ledger.API
+
+--------------------------------------------------------------------------------
+-- Translation from Shelley to Allegra
+--
+-- The instances below are needed by the consensus layer. Do not remove any of
+-- them without coordinating with consensus.
+--
+-- Please add auxiliary instances and other declarations at the bottom of this
+-- module, not in the list below so that it remains clear which instances the
+-- consensus layer needs.
+--
+-- WARNING: when a translation instance currently uses the default
+-- 'TranslationError', i.e., 'Void', it means the consensus layer relies on it
+-- being total. Do not change it!
+--------------------------------------------------------------------------------
+
+type instance PreviousEra (AllegraEra c) = ShelleyEra c
+
+-- | Currently no context is needed to translate from Shelley to Allegra.
+--
+-- Note: if context is needed, please coordinate with consensus, who will have
+-- to provide the context in the right place.
+type instance TranslationContext (AllegraEra c) = ()
+
+instance Crypto c => TranslateEra (AllegraEra c) ShelleyState where
+  translateEra _ = undefined
+
+instance Crypto c => TranslateEra (AllegraEra c) Tx where
+  type TranslationError (AllegraEra c) Tx = ()
+  translateEra _ = undefined
+
+instance Crypto c => TranslateEra (AllegraEra c) ShelleyGenesis where
+  translateEra ctxt genesis =
+    return
+      ShelleyGenesis
+        { sgSystemStart = sgSystemStart genesis,
+          sgNetworkMagic = sgNetworkMagic genesis,
+          sgNetworkId = sgNetworkId genesis,
+          sgActiveSlotsCoeff = sgActiveSlotsCoeff genesis,
+          sgSecurityParam = sgSecurityParam genesis,
+          sgEpochLength = sgEpochLength genesis,
+          sgSlotsPerKESPeriod = sgSlotsPerKESPeriod genesis,
+          sgMaxKESEvolutions = sgMaxKESEvolutions genesis,
+          sgSlotLength = sgSlotLength genesis,
+          sgUpdateQuorum = sgUpdateQuorum genesis,
+          sgMaxLovelaceSupply = sgMaxLovelaceSupply genesis,
+          sgProtocolParams = translateEra' ctxt (sgProtocolParams genesis),
+          sgGenDelegs =
+            Map.fromList
+              . map (bimap (translateEra' ctxt) (translateEra' ctxt))
+              . Map.toList
+              . sgGenDelegs
+              $ genesis,
+          sgInitialFunds = Map.mapKeys (translateEra' ctxt) $ sgInitialFunds genesis,
+          sgStaking = translateEra' ctxt $ sgStaking genesis
+        }
+
+--------------------------------------------------------------------------------
+-- The following instances are also needed by consensus, but only temporarily.
+-- These types should not be parameterised by the era but by the crypto. When
+-- this has changed (CAD-2000), the instances below can be removed.
```

Comment:
@JaredCorduan You could say that CAD-2000 is blocking this PR and this PR is blocking the [next step of the integration in consensus](https://github.com/input-output-hk/ouroboros-network/pull/2679) (i.e., making Allegra and Mary actual different eras, not just Shelley clones) :slightly_smiling_face:

Reply:
thanks for the heads up @mrBliss , I'm working on CAD-2000 now.

Reply:
I was hoping to finish this up today, but I'm still not there. my branch is `jc/remove-era-from-protocol`, it's down to the test code. here's hoping tomorrow brings it to an end :crossed_fingers:

Reply:
Great! I'll first rebase this PR on master and then have a look at your branch (and maybe start rebasing on that branch).

Reply:
Done: rebased this PR on master. I also created `mrBliss/era-translations2`, which is this branch rebased on `jc/remove-era-from-protocol`. That branch is already a bit out of date w.r.t. `master`, though.

@JaredCorduan Your branch allowed me to get rid of a bunch of `TranslateEra` instances, as expected :+1:.

I'll now test this out with my draft PR in consensus.

Reply:
@JaredCorduan Propagating `jc/remove-era-from-protocol` feels so right :relaxed:.

So far, I have encountered one problem: consensus needs to pass the `PParams` to `chainChecks`. Before your changes, we could extract them from the `LedgerView`, but that's no longer possible after your changes. Consensus doesn't have anything else in scope where it could get the `PParams` from. Potential solution: create a record with the fields needed by `chainChecks` (`_maxBHSize`, `_maxBBSize`, `_protocolVersion`), let `chainChecks` that record as an argument and store the record in the `LedgerView`.

Reply:
@JaredCorduan Another thing that I noticed: we have the following class in consensus:
```haskell
class ( Crypto c
      , SL.DSignable    c (OCertSignable c)
   -- , SL.DSignable    c (Hash era (TxBody era)) TODO
      , SL.KESignable   c (BHBody c)
      , SL.VRFSignable  c Seed
      ) => TPraosCrypto c
```
It would be a lot more convenient if that `era` wasn't there :slightly_smiling_face:.

Maybe we can do something similar to what you did for the block body? I.e.:

```haskell
newtype HashBBody crypto = UnsafeHashBBody {unHashBody :: (Hash crypto EraIndependentBodyHash)}
```

Reply:
ok, thank you for all the clear feedback @mrBliss ! I'm glad to hear that this is mostly what you expected :)

Reply:
Alternatively, adding `SL.DSignable (EraCrypto era) (Hash (EraCrypto era) (TxBody era))` to `ShelleyBased` would also be convenient.